### PR TITLE
Mobx dimension selector + add WMS Dimensions + TableMixin manual reigon mapping

### LINKED ISSUE #3970: Model Architecture: DimensionSelectorSection doesn't work with WMS
Style and Dimensions selector not wired for WMS catalog item in Mobx branch.

Example WMS layer - https://ppbcha.duckdns.org/geoserver/ows?service=WMS&version=1.3.0&request=GetCapabilities&

Then pick **PPB Inundation - 25m whole bay** layer.

## Mobx

mobx

## National Map

https://nationalmap.gov.au/#share=s-xRvpLL5Cl9ep3cAEWUGFfOL7wFb

natmap

### PR DESCRIPTION
**Pull request by nf-s**
### Merge StyleSelector and DimensionsSelector, and add WMS Dimensions

Fixes #3970

I have:

- merged the StyleSelector and DimensionsSelector, and created a SelectableDimensions interface.
- added WMS dimensions.
- region mapping controls for CSVCatalogItem (region column name and region type) - this is for YDYR
- fix WMS legends for styles
- Added support for styles/dimensions for multi-layer items
  - Styles dropdowns are created for each layer (if applicable)
  - Dimensions are flattened and applied to all layers (WMS spec says dimensions which don't apply for layers are ignored)

### Example map: WMS layer with multiple dimensions + style

Multi-layer with multiple styles and dimensions:
- http://ci.terria.io/mobx-dimension-selector/#clean&https://gist.githubusercontent.com/nf-s/3d77d4acb9105283226444677e011dd9/raw/290e89b66e2b2614462b33efb3628dbdd0ac7b99/wms-multidim.json
- Add `Multi-layer` from data catalogue

Styles with dimensions:
- http://ci.terria.io/mobx-dimension-selector/#share=s-8fkrb7LLTg3MVrobuXURHoAPQGv

Style:
- http://ci.terria.io/mobx-dimension-selector/#share=s-rGDJq0eKt60Var0hnJ7UqVuH1BC

### Example map: CSV with regions

http://ci.terria.io/mobx-dimension-selector/#share=s-zxsbPyPueLSXeOasPfLl0wSl6kX

### Checklist

-   [x] Delete StyleSelector test
-   [x] add DimensionSelector test for WMS with Dimensions and styling
-   [x] add DimensionSelector test for CSV with manual region mapping enabled
-   [x] Replace select sass with StyledComponents
-   [x] Maybe fix select size
-  [x] fix multi-layer WMS
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Models/CsvCatalogItem.ts [hunk 1]
```diff
@@ -140,6 +141,82 @@ export default class CsvCatalogItem extends TableMixin(
       );
     }
   }
+
+  @computed
+  get selectableDimensions() {
```
Comment (nf-s) [inline lib/Models/CsvCatalogItem.ts]: This adds dimensions for changing region mapping for YDYR (selecting region column and region type).
I have added an `enableManualRegionMapping` CSVCatalogItemTrait which enables this.

csv-dimensions

**Comment by nf-s** [comment]
@steve9164 and @rowanwins 
What do you think of this?
The main files of interest are `lib/Models/SelectableDimensions.ts` and `lib/ModelMixins/TableMixin.ts`.
I have followed the API Rowan made for the StyleSelector.
I could also replace the `ShadowSection` with this.

File: lib/Models/SelectableDimensions.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+export interface DimensionOption {
```
Comment (nf-s) [inline lib/Models/SelectableDimensions.ts:1]: This file contains the interface

File: lib/ModelMixins/TableMixin.ts [hunk 3]
```diff
@@ -241,34 +245,36 @@ export default function TableMixin<T extends Constructor<Model<TableTraits>>>(
     }
 
     @computed
```
Comment (nf-s) [inline lib/ModelMixins/TableMixin.ts:252]: This is where I have replaced the `styleSelector` with `selectableDimensions`

File: lib/Models/SelectableDimensions.ts (hunk 2 above)
Comment (nf-s) [reply lib/Models/SelectableDimensions.ts:1]: Should this be a Mixin instead?

Comment (tephenavies) [reply lib/Models/SelectableDimensions.ts:1]: There are instances where we've used interfaces (such as `Mappable`) and instances we've used Mixins. I think an interface works fine here.

**Comment by nf-s** [comment]
This seems to work now - but the WMS style/dimensions selector only works for single-layer wms.
I have started to add multi-layer support but it is a bit more tricky than i thought.
Maybe it would be worth creating a separate issue for that

**Comment by soyarsauce** [comment]
Good work porting the tests!

**Comment by nf-s** [comment]
> Good work porting the tests!

Thanks, sorry I didn't include your cat sample data!

One day I'll add that back in

**Comment by nf-s** [comment]
I'm sorry for the PR, I should have split it in 3 ❤️

File: lib/ReactViews/Workbench/Controls/dimension-selector-section.scss [hunk 4]
```diff
@@ -1,12 +1,46 @@
...
+    position: absolute;
+    fill: $text-light;
+    width: 10px;
+    top: 0;
+    height: 40px;
+    right: 10px;
+    pointer-events: none;
+  }
+  select {
+    background: rgba(250, 250, 250, 0.2);
```
Comment (soyarsauce) [inline lib/ReactViews/Workbench/Controls/dimension-selector-section.scss]: Will wait for your styled-components overhaul of this, but paying special attention to this to use theme colours as this will be the result if applied to another map like nsw in this form:
image

Comment (soyarsauce) [reply lib/ReactViews/Workbench/Controls/dimension-selector-section.scss]: And doubly acknowledging that it also isn't perfect in `mobx` in its un-merged StyleSelector/DimensionsSelector form

**Comment by soyarsauce** [comment]
@nf-s can you look into ci failure?

**Comment by nf-s** [comment]
I think I might leave this for the moment, as this will definitely cause breaking changes for custom catalog items. I am happy to revisit when we have a bit more time.

**Comment by nf-s** [comment]
I found a problem :(
Need to add dimension values to GetFeatureInfo

File: test/ReactViews/DimensionSelectorSectionSpec.tsx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+import CommonStrata from "../../lib/Models/CommonStrata";
+import { runInAction } from "mobx";
+import SelectableDimensions from "../../lib/Models/SelectableDimensions";
+import Select from "../../lib/Styled/Select";
+import ViewState from "../../lib/ReactViewModels/ViewState";
+import { ThemeProvider } from "styled-components";
+import { terriaTheme } from "../../lib/ReactViews/StandardUserInterface/StandardTheme";
+import CsvCatalogItem from "../../lib/Models/CsvCatalogItem";
+import loadText from "../../lib/Core/loadText";
+// import CsvCatalogItem from "../../lib/Models/CsvCatalogItem";
```
Comment (nf-s) [inline test/ReactViews/DimensionSelectorSectionSpec.tsx]: i need to remove this

File: lib/ReactViews/Workbench/Controls/dimension-selector-section.scss [hunk 6]
```diff
@@ -1,7 +1,8 @@
 @import "~terriajs-variables";
 
 .dimension-selector {
-  font-size: $font-size-mid-small;
+  font-size: 0.9rem;
+  margin-top: 10px;
   label {
     margin-right: $padding-small;
   }
```
Comment (nf-s) [inline lib/ReactViews/Workbench/Controls/dimension-selector-section.scss]: make this styled-components

File: lib/Models/WebMapServiceCatalogItem.ts [hunk 7]
```diff
@@ -456,8 +505,9 @@ class DiffStratum extends LoadableStratum(WebMapServiceCatalogItemTraits) {
   }
 
   @computed
-  get disableStyleSelector() {
-    return this.catalogItem.isShowingDiff;
+  get availableDiffStyles() {
+    // Currently only NDVI
+    return ["NDVI"];
```
Comment (nf-s) [inline lib/Models/WebMapServiceCatalogItem.ts]: check if this is still needed

File: lib/ModelMixins/TableMixin.ts [hunk 8]
```diff
@@ -259,36 +252,133 @@ export default function TableMixin<T extends Constructor<Model<TableTraits>>>(
...
-        get activeStyleId(): string | undefined {
+        get selectedId(): string | undefined {
           return tableModel.activeStyle;
         },
-        chooseActiveStyle(stratumId: string, styleId: string) {
+        setDimensionValue(stratumId: string, styleId: string) {
           tableModel.setTrait(stratumId, "activeStyle", styleId);
         }
       };
     }
```
Comment (nf-s) [inline lib/ModelMixins/TableMixin.ts]: make this consistent with the other `SelectableDimensions`

**Comment by rowanwins** [comment]
Can confirm it all works 🎉 

However
> merged the StyleSelector and DimensionsSelector, and created a SelectableDimensions interface.

I think this is the part of this PR that concerns me the most. In my mind there is a bit of a distinction between what those things do, changing styles being about visual presentation, changing dimensions being about changing what data is actually displayed.

We can see the impact of this in terms of the UI handling compared to what we got in master for multi-dimensional data, we now get a list of dimensions with no distinction between them. At the very least the "Styles" selector ought to be above the legend image I think (see images below).

In summary I think I'm have mixed feelings about whether this approach simplifies things or complicates them. Dimensions are a pretty unique concept to WMS & WCS, it feels a little like we're shoe-horning the styles and some of the table stuff into it. I'm just not sure if we're painting ourselves into a corner that'll make it harder for us in the long term...

Maybe it can work and I'm being overly cautious. I vote getting a second opinion, I nominate @steve9164 

### In Mobx
Mobx

### In Master
InMaster

**Comment by nf-s** [comment]
Yep I totally understand that, the reason I merged them is that the code was identical - there is also the `ShadowSection` which I haven't merged into this yet.
I can refactor the `DimensionsSelectorSection` so that you pass an array of `SelectableDimensions` instead of a catalog item, then you could keep two instances of it in the workbench - one for styles and one for everything else.

But I'm not sure I agree that this is specific to WMS/WCS, the main point of this for me was to make it easy to create arbitrary dropdowns for catalog items - some examples I have made for other projects:

- Geojson filter dropdowns, to filter features by properties
- A geojson intersect layer, which has dropdowns to select layers to intersect
- The region mapping stuff

**Comment by soyarsauce** [comment]
Please check if it closes https://github.com/TerriaJS/terriajs/issues/4446

**Comment by nf-s** [comment]
This closes #4446 :
image

http://ci.terria.io/mobx-dimension-selector/#clean&https://tiles.terria.io/static/terria-cube-testing-2020-05-19/australia.json

Catalog item:

image

**Comment by soyarsauce** [comment]
Should also address https://github.com/TerriaJS/terriajs/issues/4513

**Comment by tephenavies** [comment]
@nf-s Can you have a look at these broken tests? https://travis-ci.org/github/TerriaJS/terriajs/jobs/714717648#L2554

**Comment by nf-s** [comment]
@steve9164 I fixed the tests!
Let me know if you need anything else done

**Comment by soyarsauce** [comment]
Merging this should also close https://github.com/TerriaJS/terriajs/issues/4446 & subsequently https://github.com/terriajs/terriajs/issues/4390

@steve9164 incase you missed Nick's comment above

**Review by tephenavies — APPROVED** [review]

**Comment by tephenavies** [comment]
@nf-s I think this PR needs a nextra couple of WMS tests.
